399 added alternative implementation to solve #339 issue by introduci…

tomasherman opened the pull request:
…ng Distribution based Timer implementation

hanny24 reviewed (commented):
I like this better, although I still don't really understand the reasoning behind it. What is the benefit of providing a common interface for the user? (I'm not really

I would argue that this API is not easily discoverable. What I mean by that: as a user I would look into `metricsFactory.distribution` first if I wanted to use any distribution-like metric. We could still provide common interface if there were a good use case. Something like `TimerLike` trait or even a typeclass. I'm not even sure what is better as I don't have a use case for either right now.

Those are my 2¢. I'm not going to block you if you decide to go ahead with this anyway. I just trying to come up with the cleanest API possible:)

tomasherman commented on code/http4s/src/main/scala/com/avast/datadog4s/extension/http4s/DatadogMetricsOps.scala:
```diff
@@ -12,9 +12,15 @@ object DatadogMetricsOps {
   val defaultClassifierTags: ClassifierTags = classifier => List(Tag.of("classifier", classifier))
 
-  def make[F[_]](metricFactory: MetricFactory[F], classifierTags: ClassifierTags = defaultClassifierTags)(implicit
+  def make[F[_]](
```
i thought about it, but im not sure it's really worth it at this point ... it's just two parameters but if you want to create PR i would probably approve it :D

hanny24 replied:
Well...you could have avoided breaking change.

hanny24 replied:
In other words...I think it's worth it. It's not much work, and the resulting API is much more readable.

tomasherman replied:
you mean binary breaking change? because code change should not be there

hanny24 replied:
Yes, I meant binary.

tomasherman commented:
> Those are my 2¢. I'm not going to block you if you decide to go ahead with this anyway. I just trying to come up with the cleanest API possible:)

And i appreciate that you taking the time to review it over the weekend :) 

> as a user I would look into metricsFactory.distribution first if I wanted to use any distribution-like metric.

That's interesting. The way i look at it, if i'm a user and looking for timer, trying `.timer` is natural to me. And then the api forces me to make decision how the timer should behave (as histogram or as a distribution) (of course it means there will be scaladoc which i intend to write) ... For me the `Timer` is kind of a higher level object and is kind of 'above' the datadog metrics ... in other words, distribution and histogram are building blocks for higher level abstraction of `Timer` to me ...

hanny24 commented:
> That's interesting. The way i look at it, if i'm a user and looking for timer, trying `.timer` is natural to me. And then the api forces me to make decision how the timer should behave (as histogram or as a distribution) (of course it means there will be scaladoc which i intend to write) ... For me the `Timer` is kind of a higher level object and is kind of 'above' the datadog metrics ... in other words, distribution and histogram are building blocks for higher level abstraction of `Timer` to me ...

If it were up to me, I would avoid any high level abstraction at all. I believe the library should really be deadly stupid as it should add as little mental overhead as possible. And any custom abstractions make this harder.

tomasherman commented:
> If it were up to me, I would avoid any high level abstraction at all. I believe the library should really be deadly stupid as it should add as little mental overhead as possible. And any custom abstractions make this harder.

i'm not sure i want to down this thread of discussion :) the ability to use raw datadog functionality remains here, but the original goal of the library was to make life easier for programmers to get a high quality monitoring without a lot of hassle.

The readme states: _Goal of this project is to make great monitoring as easy as possible._ 

I believe Timer abstraction fits that very well. But maybe there might be a space for some refactoring, but let's not discuss that here :)

blandflakes commented:
I don't have much critical feedback as I don't want to distract those with much more vested familiarity with the library, but one question I have is whether histograms or distributions are necessarily `Timers` in the first place. I guess that's what @hanny24 is getting at with the comment about abstractions - I might want a distribution for something that _isn't_ a timer, and have to remember that it is from the timer factory (though, obviously, in the issue that inspired this, the distribution is for timing). To that point, a histogram isn't necessarily a timer either.

So I do find the argument that maybe "namespacing" them behind a `Timer` is an overly precise abstraction of what these metrics types can do convincing.

If I were to follow that line of reasoning, I could also be convinced that the factory classes gain a distribution at the same level as the other factories, and that the DatadogOps `make` function (or a builder) takes the boolean like we've done all along in this PR. That leads me back the (elaborate) "adapter" pattern which could be used to change the configuration using polymorphism - DatadogOps doesn't know anything about the switch, it just needs a factory that can hand it "timers", and the implementation of that adapter chooses histograms or distributions, but that feels like a lot of architecture astronomy - having the DataDogOps parameterize or build this makes sense to me.

tomasherman commented:
@blandflakes thank you for chiming in, i definitely appreciate it. I'm not sure i follow 100% (i sometimes struggle following what i like to call _fancy english_ :D ). But if i understand correctly, what you are proposing is IMHO already in place. There are metric types/factories for `histogram`s and `distribution`s, as can be seen [here](https://github.com/avast/datadog4s/blob/master/code/api/src/main/scala/com/avast/datadog4s/api/MetricFactory.scala). Please let me know if this is just a case of misundersting on your side or mine and you had something else in mind.

So my point is that the users still have access to the raw datadog functionality. But generally, we use `Timer[F]` a lot and it's been proven very useful abstraction so in persuit of giving people tools it's been also added to this library.

In the meanwhile i've added the build to apease @hanny24, as i've realized that he is probably right. Having a

blandflakes commented:
@tomasherman I think you understood perfectly - the fault is mine - I sort of meandered through the thought process arriving at a "what we have here is _probably_ the best approach".

I appreciate the explanation that the distribution is in fact exposed the way I was thinking, and that timers are an abstraction backed by either a histogram or a distribution - I had missed that detail.

hanny24 commented on code/http4s/src/main/scala/com/avast/datadog4s/extension/http4s/MetricsOpsBuilder.scala:
```diff
@@ -0,0 +1,39 @@
+import org.http4s.metrics.MetricsOps
+import cats.syntax.functor._
+
+private[http4s] final class MetricsOpsBuilder[F[_]: Sync](metricFactory: MetricFactory[F]) {
```
I think you meant to make the constructor private and not the whole class, didn't you?

tomasherman replied:
🤦 yep

hanny24 commented on code/http4s/src/main/scala/com/avast/datadog4s/extension/http4s/MetricsOpsBuilder.scala:
```diff
@@ -0,0 +1,39 @@
+import cats.syntax.functor._
+
+private[http4s] final class MetricsOpsBuilder[F[_]: Sync](metricFactory: MetricFactory[F]) {
+  private[this] var distributionBasedTimers: Boolean = false
```
Nicer solution would be to use custom copy method as seen for example [here](https://github.com/http4s/http4s/blob/b2892d3c61336d532c368c2490d4d0d7f91cd72b/blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClientBuilder.scala)

tomasherman replied:
fixed

tomasherman commented:
Ok, so i think i'm done with this. I added some scaladoc which hopefully explains difference between the two timers. I created new issue for reworking the html documentation for the datadog4s website itself as it's a little stale so i will fix that one as part of that issue. Other than that, i think we are good to merge

hanny24 commented on code/http4s/src/main/scala/com/avast/datadog4s/extension/http4s/MetricsOpsBuilder.scala:
```diff
@@ -0,0 +1,48 @@
+import com.avast.datadog4s.extension.http4s.impl.{ ActiveConnections, DefaultMetricsOps }
+import org.http4s.metrics.MetricsOps
+
+final case class MetricsOpsBuilder[F[_]: Sync] private (
```
nitpick: using the case class is no optimal as it exposes `copy` method. But I won't block you on this.

hanny24 commented on code/statsd/src/main/scala/com/avast/datadog4s/statsd/metric/timer/DistributionTimer.scala:
```diff
@@ -0,0 +1,22 @@
+
+import java.time.Duration
+
+class DistributionTimer[F[_]: Sync](
```
private?

hanny24 commented on code/statsd/src/main/scala/com/avast/datadog4s/statsd/metric/timer/HistogramTimer.scala:
```diff
@@ -0,0 +1,22 @@
+
+import java.time.Duration
+
+class HistogramTimer[F[_]: Sync](
```
private?

hanny24 reviewed (approved):
A few more comments, threat them as you wish.

tomasherman commented:
well mergify takes no prisoners :D

hanny24 commented:
And not even squashing commits:(

tomasherman commented:
yeah, i wonder if there is a setting for that ...

tomasherman commented:
i think i fixed it manually, hopefully i haven't fucked it up
